pkg/metrics_store/metrics_writer.go

```diff
@@ -82,20 +87,59 @@ func (m MetricsWriter) WriteAll(w io.Writer) error {
 	return nil
 }
 
-// SanitizeHeaders removes duplicate headers from the given MetricsWriterList for the same family (generated through CRS).
-// These are expected to be consecutive since G** resolution generates groups of similar metrics with same headers before moving onto the next G** spec in the CRS configuration.
-func SanitizeHeaders(writers MetricsWriterList) MetricsWriterList {
+// SanitizeHeaders sanitizes the headers of the given MetricsWriterList.
+func SanitizeHeaders(contentType string, writers MetricsWriterList) MetricsWriterList {
 	var lastHeader string
 	for _, writer := range writers {
 		if len(writer.stores) > 0 {
 			for i, header := range writer.stores[0].headers {
+
+				// Nothing to sanitize.
+				if len(header) == 0 {
+					continue
+				}
+
+				// Removes duplicate headers from the given MetricsWriterList for the same family (generated through CRS).
+				// These are expected to be consecutive since G** resolution generates groups of similar metrics with same headers before moving onto the next G** spec in the CRS configuration.
+				// Skip this step if we encounter a repeated header, as it will be removed.
+				if header != lastHeader && strings.HasPrefix(header, "# HELP") {
+
+					// If the requested content type was proto-based, replace "info" and "statesets" with "gauge", as they are not recognized by Prometheus' protobuf machinery.
+					if strings.HasPrefix(contentType, expfmt.ProtoType) {
+						infoTypeString := metric.Info.String()
+						stateSetTypeString := metric.StateSet.String()
+						if strings.HasSuffix(header, infoTypeString) {
+							header = header[:len(header)-len(infoTypeString)] + string(metric.Gauge)
+							writer.stores[0].headers[i] = header
+						}
+						if strings.HasSuffix(header, stateSetTypeString) {
+							header = header[:len(header)-len(stateSetTypeString)] + string(metric.Gauge)
+							writer.stores[0].headers[i] = header
+						}
+					}
+				}
+
+				// Nullify duplicate headers after the sanitization to not miss out on any new candidates.
 				if header == lastHeader {
 					writer.stores[0].headers[i] = ""
-				} else {
-					lastHeader = header
 				}
+
+				// Update the last header.
+				lastHeader = header
 			}
 		}
 	}
+
+	// Remove all empty headers.
+	for _, writer := range writers {
```

Comment:
This sounds like something we should optimize by merging it into the loop above (at least the writers to avoid looping through that twice)